### PR TITLE
feat: nightfall node dev branch

### PR DESCRIPTION
**Pull request by imagobea**
## What does this implement/fix? Explain your changes.

With this PR, the Proposer app is no longer needed. The optimist itself can perform this role by calling the `proposer/register` endpoint once the optimist is online. 

**Important changes in this PR**

- Protects almost all optimist routes with an API Key (#1199, #1255)
- Handles proposer related transactions autonomously, using `PROPOSER_KEY` to sign transactions (#1200, #1225)
- Adds Swagger js docs to Optimist API (#1213)
- Updates `nf3`
- Refactors tests: note that one `optimist-sync` test was discontinued, but will be restored when tackling challenger

_Note that this PR is only the beginning towards getting rid of `nf3 Proposer` functions. Not all tests were updated to discontinue this practice. Similarly, proposer app was kept for now to facilitate launching proposer in ping-pong test._

## Does this close any currently open issues?

See related Epic on ZenHub

## What commands can I run to test the change? 

```bash
# Note the params and update as appropriate
curl -k -X POST -H "X-APP-TOKEN: 0ce4fee0-c765-43d6-973c-d404bfdde2e9" -d "url=[optimist-url]&stake=999999999&fee=10" http://localhost:8081/proposer/register
```

For further details, please refer to each PR.

## Any other comments?

n/a

### CODE REVIEW

File: nightfall-optimist/src/utils/auth.mjs [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import config from 'config';
+import crypto from 'crypto';
+
+const environment = config.ENVIRONMENTS[process.env.ENVIRONMENT] || config.ENVIRONMENTS.localhost;
+
+export default function auth(req, res, next) {
+  const token = req.headers['x-app-token'];
+  if (token === crypto.createHash('sha256').update(environment.PROPOSER_KEY).digest('hex'))
```
Comment (druiz0992) [inline nightfall-optimist/src/utils/auth.mjs:8]: What happens if PROPOSER_KEY changes? Or there is no associated proposer key because this device is just a challenger? Or its undefined?
I think this is a source of problems. Change this to a fix value please

File: nightfall-optimist/src/app.mjs [hunk 2]
```diff
@@ -12,8 +19,18 @@ import {
   debug,
 } from './routes/index.mjs';
 
+const spec = swaggerJsDoc(options);
+const environment = config.ENVIRONMENTS[process.env.ENVIRONMENT] || config.ENVIRONMENTS.localhost;
+logger.debug(`********* Optimist environment vars ${JSON.stringify(environment)}`); // TODO review logs
+
 const app = express();
 
+const ethPrivateKey = environment.PROPOSER_KEY;
```
Comment (druiz0992) [inline nightfall-optimist/src/app.mjs:28]: Why are we reading proposer erh address and private key?

File: nightfall-optimist/src/services/block-assembler.mjs [hunk 3]
```diff
@@ -117,49 +126,56 @@ export async function conditionalMakeBlock(proposer) {
...
           )
           .encodeABI();
+        const { address } = web3.eth.accounts.privateKeyToAccount(ethPrivateKey);
+        const blockStake = await stateContractInstance.methods.getBlockStake().call();
+        const tx = {
+          from: address,
+          to: stateContractAddress,
+          data: txDataToSign,
+          value: blockStake,
+          gas: 8000000,
```
Comment (druiz0992) [inline nightfall-optimist/src/services/block-assembler.mjs:142]: Fix this constant. There should be a gas estimation and gas price as we have always had

File: nightfall-optimist/src/routes/block.mjs [hunk 4]
```diff
@@ -26,68 +44,66 @@ router.post('/check', async (req, res, next) => {
...
   try {
-    logger.debug(`block make-now endpoint received GET`);
     setMakeNow();
     res.send('Making short block');
   } catch (err) {
     next(err);
   }
 });
 
-router.get('/transaction-hash/:transactionHash', async (req, res, next) => {
```
Comment (druiz0992) [inline nightfall-optimist/src/routes/block.mjs:39]: Why is this endpoint removed?

File: nightfall-optimist/src/routes/block.mjs [hunk 5]
```diff
@@ -26,68 +44,66 @@ router.post('/check', async (req, res, next) => {
...
-      res.json({ block, transactions, index });
-    } else {
-      res.json({ block: null, transactions: null, index: null });
-    }
-  } catch (err) {
-    next(err);
-  }
-});
-
-router.get('/root/:root', async (req, res, next) => {
```
Comment (druiz0992) [inline nightfall-optimist/src/routes/block.mjs:60]: Why is this removed?

File: nightfall-optimist/src/services/database.mjs [hunk 6]
```diff
@@ -123,18 +123,6 @@ export async function getBlockByBlockHash(blockHash) {
   return db.collection(SUBMITTED_BLOCKS_COLLECTION).findOne(query);
 }
 
-/**
-function to get a block by root, if you know the root of the block. This
-is useful for nightfall-client to establish the layer block number containing
-a given (historic) root.
-*/
-export async function getBlockByRoot(root) {
```
Comment (druiz0992) [inline nightfall-optimist/src/services/database.mjs:131]: Why is it removed?

File: nightfall-optimist/src/app.mjs [hunk 7]
```diff
@@ -12,8 +19,18 @@ import {
   debug,
 } from './routes/index.mjs';
 
+const spec = swaggerJsDoc(options);
+const environment = config.ENVIRONMENTS[process.env.ENVIRONMENT] || config.ENVIRONMENTS.localhost;
+logger.debug(`********* Optimist environment vars ${JSON.stringify(environment)}`); // TODO review logs
```
Comment (druiz0992) [inline nightfall-optimist/src/app.mjs:24]: We cant print this

File: nightfall-optimist/src/services/block-assembler.mjs [hunk 8]
```diff
@@ -117,49 +126,56 @@ export async function conditionalMakeBlock(proposer) {
...
-        const unsignedProposeBlockTransaction = await (
-          await waitForContract(STATE_CONTRACT_NAME)
-        ).methods
+        const txDataToSign = await stateContractInstance.methods
           .proposeBlock(
             Block.buildSolidityStruct(block),
             transactions.map(t => Transaction.buildSolidityStruct(t)),
           )
           .encodeABI();
+        const { address } = web3.eth.accounts.privateKeyToAccount(ethPrivateKey);
```
Comment (druiz0992) [inline nightfall-optimist/src/services/block-assembler.mjs:135]: if we are fixing the proposer private key to the one in config, why do we need to compute the Ethereum address in every block? Or why cant we pass it as a parameter as well?

File: nightfall-optimist/src/app.mjs (hunk 7 above)
Comment (imagobea) [reply nightfall-optimist/src/app.mjs:24]: Have removed this [on the next PR](https://github.com/EYBlockchain/nightfall_3/pull/1225/files#diff-06bcbf60d087c9b88e8acb417f75e2b3c131d716431e5e36e4aa5dbfee1bdf61)

File: nightfall-optimist/src/app.mjs (hunk 2 above)
Comment (imagobea) [reply nightfall-optimist/src/app.mjs:28]: The private key is needed for signing transactions.
The address is part of the transaction. Also, most of the contracts can filter results based on address.

Is this what you meant, or do you have other concerns, ie:

- Storing key in memory
- Similar to `API key`, changing private keys

File: nightfall-optimist/src/services/block-assembler.mjs (hunk 8 above)
Comment (imagobea) [reply nightfall-optimist/src/services/block-assembler.mjs:173]: I will think of the best way to improve this depending on your answers above re Eth address and private key :pray: Thanks

File: nightfall-optimist/src/app.mjs (hunk 2 above)
Comment (signorecello) [reply nightfall-optimist/src/app.mjs:28]: We could derive an eth address from the private key, is that your proposal @druiz0992 ? Sorry I don't understand what you mean

File: nightfall-optimist/src/routes/block.mjs (hunk 4 above)
Comment (signorecello) [reply nightfall-optimist/src/routes/block.mjs:39]: It is unused

File: nightfall-optimist/src/routes/block.mjs (hunk 5 above)
Comment (signorecello) [reply nightfall-optimist/src/routes/block.mjs:60]: Also unused

File: nightfall-optimist/src/services/database.mjs (hunk 6 above)
Comment (signorecello) [reply nightfall-optimist/src/services/database.mjs:131]: Also unused

File: nightfall-optimist/src/app.mjs (hunk 7 above)
Comment (signorecello) [reply nightfall-optimist/src/app.mjs:24]: This is a draft @druiz0992 we will clean it up once it's ready

File: nightfall-optimist/src/utils/auth.mjs (hunk 1 above)
Comment (signorecello) [reply nightfall-optimist/src/utils/auth.mjs:8]: If the key changes, you also change it on the HTTP header. I agree that if the device is just a challenger, there would not be a `PROPOSER_KEY` but a `CHALLENGER_KEY`, so I agree the naming should change to something like `PRIVATE_KEY`. As for being undefined, an empty key will fail, as the tests confirm.

We don't necessarily think this could be a source of problems, but we can pass yet another environment variable if you so prefer

Comment (druiz0992) [reply nightfall-optimist/src/utils/auth.mjs:8]: I do prefer having a separate env. My reason is that we won’t have a proposer always. The second, is that for usability during testing, i find it more convenient  to have the same token regardless of optimist were connecting

File: nightfall-optimist/src/app.mjs (hunk 2 above)
Comment (druiz0992) [reply nightfall-optimist/src/app.mjs:28]: This question is just to understand the reason. I got the impression from other files that you are retrieving key and address information from env variables, so I don’t understand why you also copy here

File: nightfall-optimist/src/app.mjs (hunk 7 above)
Comment (druiz0992) [reply nightfall-optimist/src/app.mjs:24]: Just making sure we don’t forget

File: nightfall-optimist/src/routes/block.mjs (hunk 4 above)
Comment (druiz0992) [reply nightfall-optimist/src/routes/block.mjs:39]: Aren’t these functions useful. Maybe not being used now, but they could. I have not knowledge to confirm, but I wouldn’t remove it without talking to the one that added them in the first place.

File: nightfall-optimist/src/services/block-assembler.mjs (hunk 8 above)
Comment (druiz0992) [reply nightfall-optimist/src/services/block-assembler.mjs:173]: I’m talking about efficiency point of view. I am just wondering why we need to compute something that we have already computed in app.mjs

File: nightfall-optimist/src/routes/block.mjs (hunk 4 above)
Comment (imagobea) [reply nightfall-optimist/src/routes/block.mjs:39]: Perhaps we can ask on Slack about the 2 endpoints

File: nightfall-optimist/src/utils/auth.mjs (hunk 1 above)
Comment (imagobea) [reply nightfall-optimist/src/utils/auth.mjs:8]: Now addressed in #1255

File: nightfall-optimist/src/routes/block.mjs (hunk 4 above)
Comment (imagobea) [reply nightfall-optimist/src/routes/block.mjs:39]: Resolving the conversation since people seem to be okay with these removals

File: nightfall-optimist/src/services/block-assembler.mjs (hunk 3 above)
Comment (imagobea) [reply nightfall-optimist/src/services/block-assembler.mjs:185]: Addressed in #1262 
FYI I closed the previous PR for gas estimation and applied the changes here: put gas functions in common-files, preserving `gasPrice`.

File: nightfall-optimist/src/app.mjs (hunk 2 above)
Comment (imagobea) [reply nightfall-optimist/src/app.mjs:26]: Addressed in #1262

File: common-files/utils/gas.mjs [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+ *
+ * @async
+ * @function estimateGas
+ * @param {object} tx A tx object
+ * @param {Web3} web3
+ * @param {number} gasDefault Has default
+ * @param {number} gasMultiplier Buffer to apply to estimated gas - Has default
+ * @returns {Promise<number>}
+ */
+export async function estimateGas(tx, web3, gasDefault = 4000000, gasMultiplier = 2) {
```
Comment (druiz0992) [inline common-files/utils/gas.mjs:21]: can you check `common-files/utils/web3.mjs`? there is an internal function called `estimateGas`. Would you mind deleting that and using yours instead?

File: common-files/utils/gas.mjs [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+  } catch (error) {
+    try {
+      proposedGasPrice = Number(await web3.eth.getGasPrice());
+      logger.debug({ msg: 'Gas endpoint failed, web3 gas price', proposedGasPrice });
+    } catch (err) {
+      proposedGasPrice = gasPriceDefault;
+      logger.debug({ msg: 'Gas price estimation failed, use default', proposedGasPrice });
+    }
+  }
+  return Math.ceil(proposedGasPrice * gasPriceMultiplier);
```
Comment (druiz0992) [inline common-files/utils/gas.mjs:68]: why do we need a multiplier for gasPrice?

File: common-files/utils/gas.mjs [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+ *
+ * @async
+ * @function estimateGasPrice
+ * @param {string} gasEstimateEndpoint URL for querying gas prices
+ * @param {Web3} web3
+ * @param {number} gasPriceDefault Has default
+ * @param {number} gasPriceMultiplier Buffer to apply to estimated gas price - Has default
+ * @returns {Promise<number>}
+ */
+export async function estimateGasPrice(
```
Comment (druiz0992) [inline common-files/utils/gas.mjs:46]: Same as above. Can you take a look at `web3.mjs` and substitute default gasPrice value by your function?

File: nightfall-optimist/src/services/block-assembler.mjs [hunk 12]
```diff
@@ -18,11 +18,13 @@ import {
 } from './database.mjs';
 import Block from '../classes/block.mjs';
 import { Transaction } from '../classes/index.mjs';
-import {
-  increaseProposerWsFailed,
-  increaseProposerWsClosed,
-  increaseProposerBlockNotSent,
-} from './debug-counters.mjs';
+// import {
```
Comment (druiz0992) [inline nightfall-optimist/src/services/block-assembler.mjs:21]: ok to delete commented code.

File: nightfall-optimist/src/app.mjs [hunk 13]
```diff
@@ -12,8 +17,17 @@ import {
...
 } from './routes/index.mjs';
 
+const environment = config.ENVIRONMENTS[process.env.ENVIRONMENT] || config.ENVIRONMENTS.localhost;
+
 const app = express();
 
+const ethPrivateKey = environment.PROPOSER_KEY;
+const { address } = web3.eth.accounts.privateKeyToAccount(ethPrivateKey);
+
+app.set('ethPrivateKey', ethPrivateKey);
```
Comment (druiz0992) [inline nightfall-optimist/src/app.mjs:27]: i guess im still confused why you read `environment` and do `app.set('ethAddress')` and `app.set('ethPrivateKey'`) here, and not just read `environment` in `index.mjs`

File: common-files/utils/gas.mjs (hunk 10 above)
Comment (imagobea) [reply common-files/utils/gas.mjs:68]: @druiz0992  This was part of the original implementation. I guess someone thought that a buffer was needed (like in `estimateGas`). Let me know if you want to take any action.

Comment (druiz0992) [reply common-files/utils/gas.mjs:68]: i thought only gas was multiplied, not gas price

File: common-files/utils/gas.mjs (hunk 9 above)
Comment (imagobea) [reply common-files/utils/gas.mjs:21]: #1286

File: common-files/utils/gas.mjs (hunk 11 above)
Comment (imagobea) [reply common-files/utils/gas.mjs:46]: #1286

File: nightfall-optimist/src/event-handlers/rollback.mjs [hunk 14]
```diff
@@ -132,14 +130,9 @@ async function rollbackEventHandler(data) {
   ]);
 
   await dequeueEvent(2); // Remove an event from the stopQueue.
-  // A Rollback triggers a NewCurrentProposer event which shoudl trigger queue[0].end()
+  // A Rollback triggers a NewCurrentProposer event which should trigger queue[0].end()
```
Comment (israelboudoux) [inline nightfall-optimist/src/event-handlers/rollback.mjs:133]: These comments aren't necessary anymore.

File: nightfall-optimist/src/services/block-assembler.mjs [hunk 15]
```diff
@@ -175,53 +150,39 @@ export async function conditionalMakeBlock(proposer) {
...
-        ).methods
+        // Propose this block to the State contract
+        const txDataToSign = await stateContractInstance.methods
           .proposeBlock(
             Block.buildSolidityStruct(block),
             transactions.map(t => Transaction.buildSolidityStruct(t)),
           )
           .encodeABI();
 
-        // check that the websocket exists (it should) and its readyState is OPEN
```
Comment (israelboudoux) [inline nightfall-optimist/src/services/block-assembler.mjs:188]: :heart_eyes:

File: nightfall-optimist/src/services/debug-counters.mjs [hunk 16]
```diff
@@ -2,11 +2,11 @@ const debugCounters = {
   // CheckBlock returns a failure
   nBlockInvalid: 0,
   // Proposer WS is closed
-  proposerWsClosed: 0,
-  // Proposer WS is clossed and coulnd't be opened
-  proposerWsFailed: 0,
-  // Optimist couldnt proposed blocks because of WS
-  proposerBlockNotSent: 0,
+  // proposerWsClosed: 0,
```
Comment (israelboudoux) [inline nightfall-optimist/src/services/debug-counters.mjs:5]: We can kick these commented lines out :)

File: nightfall-optimist/src/services/transaction-sign-send.mjs [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+  // Check if web3 ws is available
+  const isListening = await web3.eth.net.isListening();
+  if (!isListening) throw new Error('Web3 ws not listening, try again later');
+
+  logger.debug('Create transaction object...');
+
+  let signedTx;
+  await nonceMutex.runExclusive(async () => {
+    // Get nonce
+    const nonce = await web3.eth.getTransactionCount(from);
```
Comment (israelboudoux) [inline nightfall-optimist/src/services/transaction-sign-send.mjs:35]: The `getTransactionCount` method returns the count of non-pending transactions, while it doesn't consider the Pending ones, I believe this could cause an issue. This might not being caught on our tests because Ganache is configured with mining block time = 1 second.

**Review by israelboudoux — CHANGES_REQUESTED** [review]
Just a few things to double-check.

File: apps/proposer/src/proposer.mjs [hunk 18]
```diff
@@ -4,126 +4,43 @@
 Module that runs up as a proposer
```
[the remark this answers is not in the document]
Comment (imagobea) [reply apps/proposer/src/proposer.mjs]: edbc0e4f

File: cli/lib/nf3.mjs [hunk 19]
```diff
@@ -931,25 +941,26 @@ class Nf3 {
...
-            stake,
-          );
-          resolve(receipt);
-        } catch (err) {
-          reject(err);
-        }
-      });
-    });
+    return res?.data;
+    // if (res.data.txDataToSign === '') return false; // already registered
```
[the remark this answers is not in the document]
Comment (imagobea) [reply cli/lib/nf3.mjs]: 07fdb42e

**Review by Westlad — CHANGES_REQUESTED** [review]
Just want to hold back on this while we clear some decisions about the deployment infrastructure, which may favour a separate Proposer (but also may not). Nothing to do with this code as such!

File: nightfall-optimist/src/services/debug-counters.mjs (hunk 16 above)
Comment (imagobea) [reply nightfall-optimist/src/services/debug-counters.mjs]: ec704738

File: nightfall-optimist/src/event-handlers/rollback.mjs (hunk 14 above)
Comment (imagobea) [reply nightfall-optimist/src/event-handlers/rollback.mjs]: fffd3a10 Clearly I didn't understand which comments were linked to what code... :sweat_smile:

File: nightfall-optimist/src/routes/proposer.mjs [hunk 20]
```diff
@@ -37,92 +38,235 @@ export function setProposer(p) {
...
+      logger.debug('Registering proposer with this Optimist instance...');
+      await setRegisteredProposerAddress(ethAddress, url); // save the registration address
 
-      /*
-        We've just registered with optimist but if we were already registered on the blockchain,
-        we should check if we're the current proposer and, if so, set things up so we start
-        making blocks immediately
-       */
+      // I we were already registered on the blockchain, check if we're the current proposer
       if (txDataToSign === '') {
```
[the remark this answers is not in the document]
Comment (imagobea) [reply nightfall-optimist/src/routes/proposer.mjs:144]: Well spotted! :boom:

File: config/default.js [hunk 21]
```diff
@@ -269,6 +270,7 @@ module.exports = {
       adversarialClientWsUrl: `wss://${process.env.ADVERSARY_CLIENT_HOST}`,
       PROPOSER_KEY: process.env.PROPOSER_KEY,
       CHALLENGER_KEY: process.env.CHALLENGER_KEY,
+      AUTH_TOKEN: process.env.AUTH_TOKEN,
```
[the remark this answers is not in the document]
Comment (imagobea) [reply config/default.js:267]: 28934ea9
Resolves (or attempts to) all comments re env vars, including documentation.
I have reviewed the `docker-compose` and the `docker-compose.deployment` files, adding (and re-organising :upside_down_face: ) some of the vars. I also perceived a different syntax in a few places so please take a look.
Re docs, I have updated `env.deployment.template` for now.
Again, penny for your thoughts!
Cheers

File: nightfall-optimist/src/routes/proposer.mjs [hunk 22]
```diff
@@ -37,92 +38,235 @@ export function setProposer(p) {
...
+    // Update proposer data
     const txDataToSign = await proposersContractInstance.methods
       .updateProposer(url, fee)
       .encodeABI();
 
-    res.json({ txDataToSign });
-    setRegisteredProposerAddress(address, url); // save the registration address and URL
+    // Sign tx
+    const proposersContractAddress = proposersContractInstance.options.address;
+    const signedTx = await createSignedTransaction(
```
[the remark this answers is not in the document]
Comment (imagobea) [reply nightfall-optimist/src/routes/proposer.mjs:225]: Since we agreed to update the proposer routes, just one instance will be left. But I will consider a refactor for the `challenger` implementation!

File: nightfall-optimist/src/routes/proposer.mjs [hunk 23]
```diff
@@ -37,92 +38,235 @@ export function setProposer(p) {
...
+      signedTx = await createSignedTransaction(
+        ethPrivateKey,
+        ethAddress,
+        proposersContractAddress,
+        txDataToSign,
+        _stake,
       );
+
+      // Submit tx
+      txsQueue.push(async () => {
```
[the remark this answers is not in the document]
Comment (imagobea) [reply nightfall-optimist/src/routes/proposer.mjs]: 5dfd1dfb

File: nightfall-optimist/src/services/transaction-sign-send.mjs (hunk 17 above)
Comment (imagobea) [reply nightfall-optimist/src/services/transaction-sign-send.mjs]: 57ca8b3e Changes
#1356 Trying to test this...

File: nightfall-optimist/src/routes/proposer.mjs [hunk 24]
```diff
@@ -37,92 +38,235 @@ export function setProposer(p) {
...
       );
       proposer.isMe = true;
-      proposer.address = address;
+      proposer.address = ethAddress;
       await enqueueEvent(() => logger.info('Start Queue'), 0); // kickstart the queue
     }
-    res.json({ txDataToSign });
+
+    const { transactionHash } = signedTx;
+    res.json({ transactionHash });
```
[the remark this answers is not in the document]
Comment (imagobea) [reply nightfall-optimist/src/routes/proposer.mjs]: b45d5588
There are still a couple of things missing, but these updates should do for now
Will review this again when doing `challenger`

File: .env.deployment.template [hunk 25]
```diff
@@ -87,7 +87,7 @@ export RABBITMQ_PORT=
 
 # The L2 Token ID that was defined in the settings ('config/default.js') to be used to pay the fees. See nightfall-deployer/README.md.
 # Required!
-FEE_L2_TOKEN_ID=
+export FEE_L2_TOKEN_ID=
```
Comment (israelboudoux) [inline .env.deployment.template:90]: :heart:

File: docker/docker-compose.deployment.yml [hunk 26]
```diff
@@ -72,18 +72,18 @@ services:
       ENVIRONMENT: ${ENVIRONMENT:?ENVIRONMENT-cannot-be-blank}
       BLOCKCHAIN_URL: ${BLOCKCHAIN_URL:?BLOCKCHAIN_URL-cannot-be-blank}
       CIRCOM_WORKER_HOST: ${CIRCOM_WORKER_HOST:-worker}
-      DEPLOY_MOCK_TOKENS: ${DEPLOY_MOCK_TOKENS?DEPLOY_MOCK_TOKENS-not-set}
+      DEPLOY_MOCK_TOKENS: ${DEPLOY_MOCK_TOKENS:?DEPLOY_MOCK_TOKENS-not-set}
```
Comment (israelboudoux) [inline docker/docker-compose.deployment.yml:75]: Thank you! :heart_eyes:

**Review by israelboudoux — APPROVED** [review]

**Comment by israelboudoux** [comment]
> Just want to hold back on this while we clear some decisions about the deployment infrastructure, which may favour a separate Proposer (but also may not). Nothing to do with this code as such!

Hi @Westlad,
Would you have an idea when can have this merged?

Thank you!
